tools/extract_cmake_functions.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2025 Raspberry Pi (Trading) Ltd.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#
+# Script to scan the Raspberry Pi Pico SDK tree searching for CMake functions
+# Outputs a tab separated file of the function:
+# name	signature	description	group
+#
+# Usage:
+#
+# tools/extract_cmake_functions.py <root of repo> [output file]
+#
+# If not specified, output file will be `pico_cmake_functions.tsv`
+
+
+import os
+import sys
+import re
+import csv
+import logging
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+scandir = sys.argv[1]
+outfile = sys.argv[2] if len(sys.argv) > 2 else 'pico_cmake_functions.tsv'
+
+CMAKE_FUNCTION_RE = re.compile(r'^\s*#(.*)((\n\s*#.*)*)\n\s*function\(([^\s]*)', re.MULTILINE)
+
+CMAKE_PICO_FUNCTIONS_RE = re.compile(r'^\s*function\((pico_[^\s\)]*)', re.MULTILINE)
+
+# Files containing internal functions that don't need to be documented publicly
+skip_files = set([
+    "pico_sdk_init.cmake",
+    "pico_utils.cmake",
+    "no_hardware.cmake",
+    "find_compiler.cmake",
+])
+
+skip_groups = set([
+    "src",  # skip the root src/CMakeLists.txt
+])
+
+# Other internal functions that don't need to be documented publicly
+allowed_missing_functions = set([
+    "pico_init_pioasm",
+    "pico_init_picotool",
+    "pico_add_platform_library",
+    "pico_get_runtime_output_directory",
+    "pico_set_printf_implementation",
+    "pico_expand_pico_platform",
+])
+
+all_functions = {}
+
+# Supported commands:
+# \brief\ <brief description, which should be included in the main description>
+# \brief_nodesc\ <brief description, which should be excluded from the main description>
+# \param\ <parameter_name> <parameter description>
+#
+# Commands in the middle of a line are not supported
+def process_commands(description, name, group, signature):
+    brief = ''
+    params = []
+    ret = ''
```

Comment:
I guess `ret` here is actually `desc`? (which might make the code easier to read)